Subject: [PATCH] [PPPOL2TP]: Add missing sock_put()

Kernel.org commits
c8fff1cf4e4e5e420c929469a09427aa37342928
ec9b6add7d81f902f6094e71f595da4a362f3348
e6358135147807351db3b7782d3e198a1bba8b62

Every skb removed from session->reorder_q needs sock_put().

---

diff --git a/drivers/net/pppol2tp.c b/drivers/net/pppol2tp.c
index e0b072d..dcd4991 100644
--- a/drivers/net/pppol2tp.c
+++ b/drivers/net/pppol2tp.c
@@ -491,6 +491,7 @@ static void pppol2tp_recv_dequeue(struct
 			       skb_queue_len(&session->reorder_q));
 			skb_unlink(skb, &session->reorder_q);
 			kfree_skb(skb);
+			sock_put(session->sock);
 			goto again;
 		}
 
@@ -1111,6 +1111,8 @@ static void pppol2tp_tunnel_closeall(struct pppol2tp_tunnel *tunnel)
 
 	for (hash = 0; hash < PPPOL2TP_HASH_SIZE; hash++) {
 		hlist_for_each_safe(walk, tmp, &tunnel->session_hlist[hash]) {
+			struct sk_buff *skb;
+
 			session = hlist_entry(walk, struct pppol2tp_session, hlist);
 
 			sk = session->sock;
@@ -1139,6 +1141,10 @@ again:
 			/* Purge any queued data */
 			skb_queue_purge(&sk->sk_receive_queue);
 			skb_queue_purge(&sk->sk_write_queue);
+			while ((skb = skb_dequeue(&session->reorder_q))) {
+				kfree_skb(skb);
+				sock_put(sk);
+			}
 
 			release_sock(sk);
 
@@ -1446,6 +1446,14 @@ static int pppol2tp_release(struct socke
 	/* Purge any queued data */
 	skb_queue_purge(&sk->sk_receive_queue);
 	skb_queue_purge(&sk->sk_write_queue);
+	if (session != NULL) {
+		struct sk_buff *skb;
+		while ((skb = skb_dequeue(&session->reorder_q))) {
+			kfree_skb(skb);
+			sock_put(sk);
+		}
+		sock_put(sk);
+	}
 
 	release_sock(sk);
 
-- 
1.7.1
